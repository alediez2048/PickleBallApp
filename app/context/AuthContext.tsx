import React, { createContext, useContext, useState, useEffect } from 'react';
import { Session, User } from '@supabase/supabase-js';
import { supabase } from '../config/supabase';
import * as authService from '../services/auth';
import { clearSecureStorage } from '../services/secureStorage';

// Define the shape of our auth context
type AuthContextType = {
  session: Session | null;
  user: User | null;
  profile: any | null;
  isLoading: boolean;
  signIn: (email: string, password: string) => Promise<{ error: any }>;
  signUp: (email: string, password: string) => Promise<{ error: any }>;
  signOut: () => Promise<void>;
  refreshProfile: () => Promise<void>;
};

// Create the context with a default value
const AuthContext = createContext<AuthContextType | undefined>(undefined);

// Provider component that wraps the app and makes auth object available to any child component that calls useAuth()
export function AuthProvider({ children }: { children: React.ReactNode }) {
  const [session, setSession] = useState<Session | null>(null);
  const [user, setUser] = useState<User | null>(null);
  const [profile, setProfile] = useState<any | null>(null);
  const [isLoading, setIsLoading] = useState<boolean>(true);

  // Listen for auth changes
  useEffect(() => {
    setIsLoading(true);

    // Get the current session
    authService.getSession().then(({ session, error }) => {
      setSession(session);
      setUser(session?.user ?? null);
      
      if (session?.user) {
        refreshProfile();
      }
      
      setIsLoading(false);
    });

    // Set up auth state listener
    const { data: { subscription } } = supabase.auth.onAuthStateChange(
      async (event, session) => {
        setSession(session);
        setUser(session?.user ?? null);
        
        if (session?.user) {
          await refreshProfile();
        } else {
          setProfile(null);
        }
        
        setIsLoading(false);
      }
    );

    // Cleanup subscription
    return () => {
      subscription.unsubscribe();
    };
  }, []);

  // Refresh user profile
  const refreshProfile = async () => {
    if (!user) return;
    
    const { profile, error } = await authService.getProfile(user.id);
    if (error) {
      console.error('Error fetching profile:', error);
    } else {
      setProfile(profile);
    }
  };

  // Sign in with email and password
  const signIn = async (email: string, password: string) => {
    setIsLoading(true);
    const { data, error } = await authService.signIn(email, password);
    setIsLoading(false);
    return { error };
  };

  // Sign up with email and password
  const signUp = async (email: string, password: string) => {
    setIsLoading(true);
    const { data, error } = await authService.signUp(email, password);
    setIsLoading(false);
    return { error };
  };

  // Sign out
  const signOut = async () => {
    setIsLoading(true);
    await authService.signOut();
    await clearSecureStorage();
    setIsLoading(false);
  };

  // The value that will be supplied to any components that use this hook
  const value = {
    session,
    user,
    profile,
    isLoading,
    signIn,
    signUp,
    signOut,
    refreshProfile,
  };

  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
}

// Custom hook that shorthands the context
export const useAuth = () => {
  const context = useContext(AuthContext);
  
  if (context === undefined) {
    throw new Error('useAuth must be used within an AuthProvider');
  }
  
  return context;
}; 