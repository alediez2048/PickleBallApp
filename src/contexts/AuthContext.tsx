import React, {
  createContext,
  useContext,
  useState,
  useEffect,
  useCallback,
} from "react";
import { router } from "expo-router";
import {
  mockApi,
  FirstTimeProfileData,
  UpdateProfileData,
} from "@/services/mockApi";
import { storage } from "@/services/storage";
import { socialAuth } from "@/services/socialAuth";
import { Alert, Platform } from "react-native";
import { MembershipPlan } from "@/types/membership";
import {
  signUpWithEmail,
  signInWithEmail,
  getSession,
  logout,
} from "@/services/authService";
import { createUserProfile } from "@/services/userService";
import AsyncStorage from "@react-native-async-storage/async-storage";

interface PaymentMethod {
  id: string;
  last4: string;
  brand: string;
  expiryMonth: string;
  expiryYear: string;
  isDefault: boolean;
}

interface UserProfile {
  id?: string;
  email?: string;
  name?: string;
  isVerified?: boolean;
  skillLevel?: string;
  profileImage?:
    | string
    | {
        uri: string;
        base64: string;
        timestamp: number;
      };
  phoneNumber?: string;
  dateOfBirth?: string;
  address?: {
    street?: string;
    city?: string;
    state?: string;
    zipCode?: string;
    country?: string;
  };
  membership?: MembershipPlan;
  paymentMethods?: PaymentMethod[];
  hasCompletedProfile?: boolean;
}

interface AuthState {
  user: UserProfile | null;
  token: string | null;
  session: any | null;
  isLoading: boolean;
}

interface AuthContextType extends AuthState {
  // Authentication methods
  signIn: (email: string, password: string) => Promise<void>;
  signUp: (email: string, password: string, name: string) => Promise<void>;
  signOut: () => Promise<void>;

  // Social authentication methods
  signInWithGoogle: () => Promise<void>;
  signInWithFacebook: () => Promise<void>;

  // Profile management methods
  updateProfile: (updates: Partial<UserProfile>) => Promise<void>;
  updateFirstTimeProfile: (data: FirstTimeProfileData) => Promise<void>;
  updateMembership: (plan: MembershipPlan) => Promise<void>;
  updatePaymentMethods: (methods: PaymentMethod[]) => Promise<void>;
  updatePaymentMethod?: (hasPaymentMethod: boolean) => Promise<void>;

  // Auth state
  isAuthenticated: boolean;
}

const AuthContext = createContext<AuthContextType | undefined>(undefined);

export function AuthProvider({ children }: { children: React.ReactNode }) {
  const [state, setState] = useState<AuthState>({
    user: null,
    token: null,
    isLoading: true,
  });

  useEffect(() => {
    // Load token and user data from storage on app start
    loadStoredAuth();
  }, []);

  useEffect(() => {
    // Only navigate when auth state changes, not on every loading state change
    if (!state.isLoading && state.token === null) {
      router.replace("/login");
    }
  }, [state.token]);

  const loadStoredAuth = async () => {
    try {
      console.log("Loading stored auth data...");

      const [token, userString] = await Promise.all([
        AsyncStorage.getItem("auth_token"),
        AsyncStorage.getItem("user"),
      ]);

      console.log("Stored auth data:", { token, userString });

      if (token && userString) {
        const user = JSON.parse(userString);

        const { data, error } = await getSession();

        if (error || !data?.session) {
          console.warn("Session expired or invalid in Supabase");
          await AsyncStorage.multiRemove(["auth_token", "user"]);
          setState({
            token: null,
            session: null,
            user: null,
            isLoading: false,
          });
          return;
        }

        console.log("Supabase session is valid");
        setState({
          token: data.session.access_token,
          session: data.session,
          user,
          isLoading: false,
        });
      } else {
        setState({ token: null, session: null, user: null, isLoading: false });
      }
    } catch (error) {
      console.error("Error loading auth data:", error);
      setState({ token: null, session: null, user: null, isLoading: false });
    }
  };

  const signUp = async (email: string, password: string, name: string) => {
    try {
      console.log("[Auth] Starting registration process", { email, name });
      setState((prev) => ({ ...prev, isLoading: true }));

      const { data, error } = await signUpWithEmail(email, password);
      if (error) throw new Error(error.message);

      const userId = data?.user?.id;
      if (userId) {
        await createUserProfile(userId, { name, email });
        setState({
          session: data.session,
          token: data.session?.access_token || null,
          user: data.user,
          isLoading: false,
        });

        // Store the auth data
        await Promise.all([
          data.session?.access_token
            ? storage.setItem("auth_token", data.session.access_token)
            : Promise.resolve(),
          storage.setItem("user", JSON.stringify(data.user)),
        ]);
        console.log("[Auth] User data stored in local storage");
      }
    } catch (error) {
      console.error("[Auth] Registration failed:", error);
      setState((prev) => ({ ...prev, isLoading: false }));
      throw error;
    }
  };

  const signIn = async (email: string, password: string) => {
    try {
      console.log("Signing in...", { email });
      setState((prev) => ({ ...prev, isLoading: true }));

      const { data, error } = await signInWithEmail(email, password);

      if (error || !data.session) {
        throw new Error(error?.message || "No session returned");
      }

      const token = data.session.access_token;
      const user = data.user;
      const session = data.session;

      console.log("Sign in successful:", { token, user });

      await Promise.all([
        AsyncStorage.setItem("auth_token", token),
        AsyncStorage.setItem("user", JSON.stringify(user)),
      ]);

      console.log("Auth data stored");

      setState({ token, session, user, isLoading: false });
    } catch (error: any) {
      console.error("Sign in error:", error.message);
      setState((prev) => ({ ...prev, isLoading: false }));
      throw error;
    }
  };

  const signOut = async () => {
    try {
      console.log("Signing out...");
      setState((prev) => ({ ...prev, isLoading: true }));

      await Promise.all([
        storage.removeItem("auth_token"),
        storage.removeItem("user"),
      ]);
      console.log("Auth data removed");

      setState({ token: null, user: null, isLoading: false });
    } catch (error) {
      console.error("Sign out error:", error);
      setState((prev) => ({ ...prev, isLoading: false }));
      throw error;
    }
  };

  const handleSocialAuthSuccess = async (
    provider: "google" | "facebook",
    token: string,
    user: { id: string; email: string; name: string; photoUrl?: string }
  ) => {
    try {
      const response = await mockApi.socialAuth({ token, provider, user });

      await Promise.all([
        storage.setItem("auth_token", response.token),
        storage.setItem("user", JSON.stringify(response.user)),
      ]);

      setState({
        token: response.token,
        user: response.user,
        isLoading: false,
      });
    } catch (error) {
      console.error(`${provider} auth error:`, error);
      Alert.alert("Error", `Failed to authenticate with ${provider}`);
    }
  };

  const signInWithGoogle = async () => {
    try {
      setState((prev) => ({ ...prev, isLoading: true }));
      const result = await socialAuth.signInWithGoogle();

      if (result.type === "success" && result.token && result.user) {
        await handleSocialAuthSuccess("google", result.token, result.user);
      } else {
        throw new Error(result.message || "Google sign in failed");
      }
    } catch (error) {
      console.error("Google sign in error:", error);
      Alert.alert("Error", "Failed to sign in with Google");
      setState((prev) => ({ ...prev, isLoading: false }));
    }
  };

  const signInWithFacebook = async () => {
    try {
      setState((prev) => ({ ...prev, isLoading: true }));
      const result = await socialAuth.signInWithFacebook();

      if (result.type === "success" && result.token && result.user) {
        await handleSocialAuthSuccess("facebook", result.token, result.user);
      } else {
        throw new Error(result.message || "Facebook sign in failed");
      }
    } catch (error) {
      console.error("Facebook sign in error:", error);
      Alert.alert("Error", "Failed to sign in with Facebook");
      setState((prev) => ({ ...prev, isLoading: false }));
    }
  };

  const updateProfile = async (updates: Partial<UserProfile>) => {
    try {
      setState((prev) => ({ ...prev, isLoading: true }));

      if (!state.user?.email) {
        throw new Error("No authenticated user");
      }

      // Map UserProfile to UpdateProfileData
      const profileUpdates: UpdateProfileData = { ...updates } as any;

      // Handle address mapping if it exists
      if (updates.address) {
        profileUpdates.address = {
          address: updates.address.street || "",
          city: updates.address.city || "",
          state: updates.address.state || "",
          zipCode: updates.address.zipCode || "",
          country: updates.address.country || "United States",
        };
      }

      const { user: updatedUser } = await mockApi.updateProfile(
        state.user.email,
        profileUpdates
      );

      // Store the updated user data
      await storage.setItem("user", JSON.stringify(updatedUser));

      setState((prev) => ({
        ...prev,
        user: updatedUser,
        isLoading: false,
      }));

      console.log("Profile updated successfully:", updatedUser);
    } catch (error) {
      console.error("Profile update error:", error);
      setState((prev) => ({ ...prev, isLoading: false }));
      throw error;
    }
  };

  const updateFirstTimeProfile = async (data: FirstTimeProfileData) => {
    try {
      console.debug("[AuthContext] Starting first time profile update", {
        platform: Platform.OS,
        hasUser: !!state.user,
        userEmail: state.user?.email,
        data,
      });

      setState((prev) => ({ ...prev, isLoading: true }));

      if (!state.user?.email) {
        console.error("[AuthContext] No authenticated user found");
        throw new Error("No authenticated user");
      }

      // Map FirstTimeProfileData to UpdateProfileData
      const profileData: UpdateProfileData = {
        skillLevel: data.skillLevel,
        displayName: data.displayName,
        phoneNumber: data.phoneNumber,
        dateOfBirth: data.dateOfBirth,
        address: {
          address: data.address.address,
          city: data.address.city,
          state: data.address.state,
          zipCode: data.address.zipCode,
          country: data.address.country,
        },
        hasCompletedProfile: true,
      };

      console.debug("[AuthContext] Calling mockApi.updateProfile", {
        email: state.user.email,
        profileData,
      });

      const { user: updatedUser } = await mockApi.updateProfile(
        state.user.email,
        profileData
      );

      console.debug("[AuthContext] Profile update API call successful", {
        updatedUser,
      });

      // After successful profile update, update the stored user data
      const userWithProfile = {
        ...updatedUser,
        hasCompletedProfile: true,
      };

      console.debug("[AuthContext] Storing updated user data");
      await storage.setItem("user", JSON.stringify(userWithProfile));

      console.debug("[AuthContext] Updating state with new user data");
      setState((prev) => ({
        ...prev,
        user: userWithProfile,
        isLoading: false,
      }));

      console.debug(
        "[AuthContext] First time profile update completed successfully"
      );
    } catch (error) {
      console.error("[AuthContext] Profile update error:", error);
      setState((prev) => ({ ...prev, isLoading: false }));
      throw error;
    }
  };

  const updateMembership = async (plan: MembershipPlan) => {
    try {
      if (!state.user) {
        throw new Error("No authenticated user");
      }

      // Update user with new membership plan
      const updatedUser = {
        ...state.user,
        membership: plan,
      };

      // Store the updated user data
      await storage.setItem("user", JSON.stringify(updatedUser));

      // Update state
      setState((prev) => ({
        ...prev,
        user: updatedUser,
      }));

      console.log("Membership updated successfully:", plan);
    } catch (error) {
      console.error("Error updating membership:", error);
      throw error;
    }
  };

  const updatePaymentMethods = async (methods: PaymentMethod[]) => {
    try {
      if (!state.user) {
        throw new Error("No authenticated user");
      }

      // Update user with new payment methods
      const updatedUser = {
        ...state.user,
        paymentMethods: methods,
      };

      // Store the updated user data
      await storage.setItem("user", JSON.stringify(updatedUser));

      // Update state
      setState((prev) => ({
        ...prev,
        user: updatedUser,
      }));

      console.log("Payment methods updated successfully:", methods);
    } catch (error) {
      console.error("Error updating payment methods:", error);
      throw error;
    }
  };

  const updatePaymentMethod = async (hasPaymentMethod: boolean) => {
    try {
      if (!state.user) {
        throw new Error("No authenticated user");
      }

      // Update user with new payment method
      const updatedUser = {
        ...state.user,
        paymentMethods: hasPaymentMethod
          ? [
              {
                id: "new_payment_method",
                last4: "XXXX",
                brand: "New",
                expiryMonth: "12",
                expiryYear: "2024",
                isDefault: true,
              },
            ]
          : [],
      };

      // Store the updated user data
      await storage.setItem("user", JSON.stringify(updatedUser));

      // Update state
      setState((prev) => ({
        ...prev,
        user: updatedUser,
      }));

      console.log(
        "Payment method updated successfully:",
        hasPaymentMethod ? "Added" : "Removed"
      );
    } catch (error) {
      console.error("Error updating payment method:", error);
      throw error;
    }
  };

  const value = {
    ...state,
    signIn,
    signUp,
    signOut,
    signInWithGoogle,
    signInWithFacebook,
    updateProfile,
    updateFirstTimeProfile,
    updateMembership,
    updatePaymentMethods,
    updatePaymentMethod,
    isAuthenticated: !!state.token,
  };

  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
}

export function useAuth() {
  const context = useContext(AuthContext);
  if (context === undefined) {
    throw new Error("useAuth must be used within an AuthProvider");
  }
  return context;
}
