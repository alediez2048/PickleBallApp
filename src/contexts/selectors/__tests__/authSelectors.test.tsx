import React from 'react';
import renderer from 'react-test-renderer';
import { AuthProvider } from '../../AuthContext';
import {
  useUser,
  useUserProfile,
  useAuthState,
  useSocialAuthState,
} from '../authSelectors';

interface MockUser {
  id: string;
  name: string;
  email: string;
  emailVerified: boolean;
  isVerified?: boolean;
}

interface MockAuthContext {
  user: MockUser | null;
  isAuthenticated: boolean;
  isLoading: boolean;
  token: string | null;
  signInWithGoogle: (() => Promise<void>) | null;
  signInWithFacebook: (() => Promise<void>) | null;
}

const mockUser: MockUser = {
  id: '1',
  name: 'Test User',
  email: 'test@example.com',
  emailVerified: true,
  isVerified: true,
};

const mockAuthContext: MockAuthContext = {
  user: mockUser,
  isAuthenticated: true,
  isLoading: false,
  token: 'mock-token',
  signInWithGoogle: () => Promise.resolve(),
  signInWithFacebook: null,
};

jest.mock('../../AuthContext', () => ({
  AuthProvider: ({ children }: { children: React.ReactNode }) => children,
  useAuth: () => mockAuthContext,
}));

// Components to test the hooks
const UserComponent: React.FC = () => {
  const user = useUser();
  return (
    <div data-testid="user-component">
      {user ? (
        <div>
          <div>ID: {user.id}</div>
          <div>Name: {user.name}</div>
          <div>Email: {user.email}</div>
          <div>Email Verified: {user.isVerified ? 'Yes' : 'No'}</div>
        </div>
      ) : (
        <div>No user found</div>
      )}
    </div>
  );
};

const UserProfileComponent: React.FC = () => {
  const profile = useUserProfile();
  return (
    <div data-testid="profile-component">
      <div>ID: {profile.id || 'undefined'}</div>
      <div>Name: {profile.name || 'undefined'}</div>
      <div>Email: {profile.email || 'undefined'}</div>
      <div>Verified: {profile.isVerified ? 'Yes' : 'No'}</div>
    </div>
  );
};

const AuthStateComponent: React.FC = () => {
  const authState = useAuthState();
  return (
    <div data-testid="auth-state-component">
      <div>Is Authenticated: {authState.isAuthenticated ? 'Yes' : 'No'}</div>
      <div>Is Loading: {authState.isLoading ? 'Yes' : 'No'}</div>
      <div>Has Token: {authState.hasToken ? 'Yes' : 'No'}</div>
      <div>Is Ready: {authState.isReady ? 'Yes' : 'No'}</div>
      <div>Is Unauthenticated: {authState.isUnauthenticated ? 'Yes' : 'No'}</div>
    </div>
  );
};

const SocialAuthStateComponent: React.FC = () => {
  const socialAuthState = useSocialAuthState();
  return (
    <div data-testid="social-auth-state-component">
      <div>Has Google Auth: {socialAuthState.hasGoogleAuth ? 'Yes' : 'No'}</div>
      <div>Has Facebook Auth: {socialAuthState.hasFacebookAuth ? 'Yes' : 'No'}</div>
    </div>
  );
};

describe('Auth Selectors', () => {
  const wrapper = (component: React.ReactElement) => (
    <AuthProvider>{component}</AuthProvider>
  );

  describe('useUser', () => {
    it('returns the current user', () => {
      mockAuthContext.user = mockUser;
      const tree = renderer.create(wrapper(<UserComponent />)).toJSON();
      expect(tree).toMatchSnapshot('authenticated user');
    });

    it('returns null when no user is authenticated', () => {
      mockAuthContext.user = null;
      const tree = renderer.create(wrapper(<UserComponent />)).toJSON();
      expect(tree).toMatchSnapshot('no user');
    });
  });

  describe('useUserProfile', () => {
    beforeEach(() => {
      mockAuthContext.user = mockUser;
    });

    it('returns formatted user profile', () => {
      const tree = renderer.create(wrapper(<UserProfileComponent />)).toJSON();
      expect(tree).toMatchSnapshot('authenticated user profile');
    });

    it('returns null values when no user is authenticated', () => {
      mockAuthContext.user = null;
      const tree = renderer.create(wrapper(<UserProfileComponent />)).toJSON();
      expect(tree).toMatchSnapshot('empty user profile');
    });
  });

  describe('useAuthState', () => {
    it('returns correct auth state when authenticated', () => {
      mockAuthContext.isAuthenticated = true;
      mockAuthContext.isLoading = false;
      mockAuthContext.token = 'mock-token';

      const tree = renderer.create(wrapper(<AuthStateComponent />)).toJSON();
      expect(tree).toMatchSnapshot('authenticated state');
    });

    it('returns correct auth state when not authenticated', () => {
      mockAuthContext.isAuthenticated = false;
      mockAuthContext.isLoading = false;
      mockAuthContext.token = null;

      const tree = renderer.create(wrapper(<AuthStateComponent />)).toJSON();
      expect(tree).toMatchSnapshot('unauthenticated state');
    });

    it('returns correct auth state while loading', () => {
      mockAuthContext.isAuthenticated = false;
      mockAuthContext.isLoading = true;
      mockAuthContext.token = null;

      const tree = renderer.create(wrapper(<AuthStateComponent />)).toJSON();
      expect(tree).toMatchSnapshot('loading state');
    });
  });

  describe('useSocialAuthState', () => {
    it('returns correct social auth state', () => {
      mockAuthContext.signInWithGoogle = () => Promise.resolve();
      mockAuthContext.signInWithFacebook = null;

      const tree = renderer.create(wrapper(<SocialAuthStateComponent />)).toJSON();
      expect(tree).toMatchSnapshot('google auth available');
    });

    it('returns false for both providers when not available', () => {
      mockAuthContext.signInWithGoogle = null;
      mockAuthContext.signInWithFacebook = null;

      const tree = renderer.create(wrapper(<SocialAuthStateComponent />)).toJSON();
      expect(tree).toMatchSnapshot('no social auth available');
    });
  });
}); 