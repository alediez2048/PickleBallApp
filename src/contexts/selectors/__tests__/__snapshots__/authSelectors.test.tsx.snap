// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth Selectors useAuthState returns correct auth state when authenticated: authenticated state 1`] = `null`;

exports[`Auth Selectors useAuthState returns correct auth state when not authenticated: unauthenticated state 1`] = `null`;

exports[`Auth Selectors useAuthState returns correct auth state while loading: loading state 1`] = `null`;

exports[`Auth Selectors useSocialAuthState returns correct social auth state: google auth available 1`] = `null`;

exports[`Auth Selectors useSocialAuthState returns false for both providers when not available: no social auth available 1`] = `null`;

exports[`Auth Selectors useUser returns null when no user is authenticated: no user 1`] = `null`;

exports[`Auth Selectors useUser returns the current user: authenticated user 1`] = `null`;

exports[`Auth Selectors useUserProfile returns formatted user profile: authenticated user profile 1`] = `null`;

exports[`Auth Selectors useUserProfile returns null values when no user is authenticated: empty user profile 1`] = `null`;
